File: ios/liveness/AAILivenessSDK-V2.0.3/Demo/LivenessSDKObjCDemo/LivenessSDKObjCDemo/AAICustomLD03Controller.m
//
//  AAICustomLD03Controller.m
//  LivenessSDKObjCDemo
//
//  Created by advance on 2022/12/7.
//

#import "AAICustomLD03Controller.h"

@interface AAICustomLD03Controller ()
{
    UILabel *_myTitleLabel;
    UIImageView *_myImgBorderView;
}
@end

@implementation AAICustomLD03Controller

- (void)livenessWrapViewDidLoad:(AAILivenessWrapView *)wrapView
{
    [super livenessWrapViewDidLoad:wrapView];
    
    // Subclass can override this method to customize the UI
//    wrapView.backgroundColor = [UIColor grayColor];
//    wrapView.roundBorderView.layer.borderColor = [UIColor redColor].CGColor;
//    wrapView.roundBorderView.layer.borderWidth = 2;
     
    // Custom corner radius and the shape of the preview area
    CGFloat cornerRadius = 0;
    wrapView.roundBorderView.layer.cornerRadius = cornerRadius;
    wrapView.configAvatarPreviewPathV2 = ^(CGRect avatarPreviewFrame, UIBezierPath * _Nonnull originRectPath, AAILivenessWrapView * _Nonnull originWrapView) {
        UIBezierPath *squarePath = [UIBezierPath bezierPathWithRoundedRect:avatarPreviewFrame cornerRadius:cornerRadius];
        [originRectPath appendPath: [squarePath bezierPathByReversingPath]];
    };

    /*
    //Custom preview area margin top
    wrapView.configAvatarPreviewMarginTop = ^CGFloat(CGRect wrapViewFrame) {
        return 64;
    };
     
    //Custom preview area width
    wrapView.configAvatarPreviewWidth = ^CGFloat(CGRect wrapViewFrame) {
        return 300;
    };
    */
    
    /*
    // Configure roundBorderView or hide it or add your own if needed
    wrapView.roundBorderView.hidden = YES;
     */
}

- (void)loadAdditionalUI
{
    [super loadAdditionalUI];
    
    // Configure your own UI
    // e.g. Remove UI controls that are not important to you.
    _stateImgView.hidden = YES;
    
    [_voiceBtn removeFromSuperview];
    
    _timeLabel.textColor = [UIColor blueColor];
    _timeLabel.layer.borderColor = [UIColor blackColor].CGColor;
    _stateLabel.textColor = [UIColor blackColor];
    
    UIImage *backBtnImg = [_backBtn imageForState:UIControlStateHighlighted];
    if (backBtnImg) {
        UIImage *newBackBtnImg = [backBtnImg imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        [_backBtn setImage:newBackBtnImg forState:UIControlStateNormal];
        _backBtn.tintColor = [UIColor redColor];
    }
    
    // Add your UI
    _myTitleLabel = [[UILabel alloc] init];
    _myTitleLabel.textColor = [UIColor greenColor];
    _myTitleLabel.textAlignment = NSTextAlignmentCenter;
    _myTitleLabel.text = self.title;
    [self.view addSubview:_myTitleLabel];
    
    _myImgBorderView = [[UIImageView alloc] init];
    _myImgBorderView.layer.borderColor = UIColor.blueColor.CGColor;
    _myImgBorderView.layer.borderWidth = 2;
    [self.view addSubview:_myImgBorderView];
}

- (void)layoutAdditionalUI
{
    [super layoutAdditionalUI];
    
    // Ajust UI
    CGFloat offsetY = 10;
    CGRect originFrame = _backBtn.frame;
    originFrame.origin.y -= offsetY;
    _backBtn.frame = originFrame;
    
    originFrame = _timeLabel.frame;
    originFrame.origin.y -= offsetY;
    _timeLabel.frame = originFrame;
    
    CGFloat sw = self.view.frame.size.width;
    CGRect titleLabelBound = CGRectMake(0, 0, (sw - CGRectGetMaxX(_backBtn.frame) * 2), _backBtn.frame.size.height);
    _myTitleLabel.bounds = titleLabelBound;
    _myTitleLabel.center = CGPointMake(sw/2, _backBtn.center.y);
    
    _myImgBorderView.frame = _roundViewFrame;
}

- (void)tapBackBtnAction
{
    // Customize the back button logic
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleDefault;
}

@end

File: ios/liveness/AAILivenessSDK-V2.0.3/Demo/LivenessSDKSwiftDemo/LivenessSDKSwiftDemo/AAICustomLD03Controller.swift
//
//  AAICustomLD03Controller.swift
//  LivenessSDKSwiftDemo
//
//  Created by advance on 2022/12/7.
//

import Foundation
import AAILiveness
import UIKit

class AAICustomLD03Controller: AAILivenessViewController {
    private var _myViewfinder: CAShapeLayer?
    private var _myTitleLabel: UILabel?
    private var _myImgBorderView: UIImageView?
    
    override func livenessWrapViewDidLoad(_ wrapView: AAILivenessWrapView) {
        super.livenessWrapViewDidLoad(wrapView)
        
        // Subclass can override this method to customize the UI
//        wrapView.backgroundColor = .gray
//        wrapView.roundBorderView.layer.borderColor = UIColor.red.cgColor;
//        wrapView.roundBorderView.layer.borderWidth = 2
        
        // Custom corner radius and the shape of the preview area
        let cornerRadius: CGFloat = 0;
        wrapView.roundBorderView.layer.cornerRadius = cornerRadius
        
        wrapView.configAvatarPreviewPathV2 = {(avatarPreviewFrame: CGRect, originRectPath: UIBezierPath, originWrapView: AAILivenessWrapView) in
            let squarePath = UIBezierPath(roundedRect: avatarPreviewFrame, cornerRadius: cornerRadius)
            originRectPath.append(squarePath.reversing())
        }
        
        /*
        //Custom preview area margin top
        wrapView.configAvatarPreviewMarginTop = {(wrapViewFrame) in
            return 64
        }
        */
        
        /*
        //Custom preview area width
        wrapView.configAvatarPreviewWidth = {(wrapViewFrame) in
            return 300
        }
        */
        
        /*
        // Configure roundBorderView or hide it or add your own if needed
        wrapView.roundBorderView.isHidden = true
         */
    }
    
    
    
    override func loadAdditionalUI() {
        super.loadAdditionalUI()
        
        // Configure your own UI
        // e.g. Remove UI controls that are not important to you.
        if let stateImgView = self.value(forKey: "_stateImgView") as? UIImageView {
            stateImgView.isHidden = true
        }
        
        if let timeLabel = self.value(forKey: "_timeLabel") as? UILabel {
            timeLabel.textColor = .blue
            timeLabel.layer.borderColor = UIColor.black.cgColor
        }
        
        if let stateLabel = self.value(forKey: "_stateLabel") as? UILabel {
            stateLabel.textColor = .black
        }
        
        if let voiceBtn = self.value(forKey: "_voiceBtn") as? UIButton {
            voiceBtn.removeFromSuperview()
        }
        
        if let backBtn = self.value(forKey: "_backBtn") as? UIButton {
            if let backBtnImg = backBtn.image(for: .normal) {
                let newBackBtnImg = backBtnImg.withRenderingMode(.alwaysTemplate)
                backBtn.setImage(newBackBtnImg, for: .normal)
                backBtn.tintColor = .red
            }
        }
        
        let myTitleLabel = UILabel()
        myTitleLabel.textColor = .green
        myTitleLabel.textAlignment = .center
        myTitleLabel.text = self.title
        self.view.addSubview(myTitleLabel)
        _myTitleLabel = myTitleLabel
        
        let imgBorderView = UIImageView()
        imgBorderView.layer.borderColor = UIColor.blue.cgColor
        imgBorderView.layer.borderWidth = 2
        self.view.addSubview(imgBorderView)
        _myImgBorderView = imgBorderView
    }
    
    override func layoutAdditionalUI() {
        super.layoutAdditionalUI()
        
        guard let backBtn = self.value(forKey: "_backBtn") as? UIButton else {
            return
        }
        
        guard let timeLabel = self.value(forKey: "_timeLabel") as? UILabel else {
            return
        }
        
        // Ajust UI
        let offsetY: CGFloat = 10
        var originFrame = backBtn.frame
        originFrame.origin.y -= offsetY;
        backBtn.frame = originFrame;
        
        originFrame = timeLabel.frame;
        originFrame.origin.y -= offsetY;
        timeLabel.frame = originFrame;
        
        let sw = self.view.frame.size.width
        
        let titleLabelBound = CGRect(x: 0, y: 0, width: (sw - backBtn.frame.maxX * 2), height: backBtn.frame.size.height)
        _myTitleLabel?.bounds = titleLabelBound
        _myTitleLabel?.center = CGPoint(x: sw/2, y: backBtn.center.y)
        
        guard let roundViewFrame = self.value(forKey: "_roundViewFrame") as? CGRect else {
            return
        }
        _myImgBorderView?.frame = roundViewFrame
    }
    
    override func tapBackBtnAction() {
        // Customize the back button logic
        self.presentingViewController?.dismiss(animated: true, completion: nil)
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .default
    }
}
